### DISCUSSION TITLE
Inherited Class Constraints

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#5681] [Language Ideas] [answered]
Today, CSharp allows method constraints to automatically inherited.  For example:
```
public abstract class C1 {
    public abstract void M1<T>() where T : class;
}

public class C2 : C1 {
    public override void M1<T>() {
        
    }
}
```

You'll notice that the ```where T : class``` constraint does not have to be restated on C2.

When working with classes, however, C# requires that constraints always be inherited.  For example:
```
public class C2<T1, T2>
   where T1 : class
   where T2 : IEnumerable<T1> {
}

public class C3<TT1, TT2, TT3> : C2<TT1, TT2>
    where TT1 : class
    where TT2 : IEnumerable<TT1>
    where TT3 : ICollection<TT1> {
}

public class C1<TTT1> : C3<TTT1, TTT1, TTT1> 
    where TTT1 : class, IEnumerable<TTT1>, ICollection<TTT1>
    {

}
```

You'll notice that there are lots of constraints that are restated.  However, the information that is restated is known (or at least deduceable) by the compiler.  The above code should be able to be simplified into the following:
```
public class C2<T1, T2>
   where T1 : class
   where T2 : IEnumerable<T1> {
}

public class C3<TT1, TT2, TT3> : C2<TT1, TT2>
    where TT3 : ICollection<TT1> {
}

public class C1<TTT1> : C3<TTT1, TTT1, TTT1>  
where TTT1 : IDictionary<long, TTT1> {

}
```

===Working through an Example===
To deduce the necessary constraints, the usage of a type parameter, only within the "inheritance" definition, should be examined, and then unioned with any explicitly stated constraints.

In walking through the above code one class at a time:
```
public class C2<T1, T2>
   where T1 : class
   where T2 : IEnumerable<T1> {
}
```
C1 does not inherit from anything.
T1 gains the explicit ```: class``` constraint
T2 gains the explicit ```IEnumerable<T1>``` constraint

```
public class C3<TT1, TT2, TT3> : C2<TT1, TT2>
    where TT3 : ICollection<TT1> {
}
```
C3 inherits from C2 so we will first look at the type parameters passed to C2.
TT1 is used as the type parameter to C2<T1, __> so it will infer the ```: class``` constraint.
TT2 is used as the type parameter to C2<__, T2> so it will infer the ```: IEnumerable<T1>``` constraint.
TT3 gains the explicit ```: ICollection<TT1>``` constraint

```
public class C1<TTT1> : C3<TTT1, TTT1, TTT1>  
where TTT1 : IDictionary<long, TTT1> {

}
```
C1 inherits from C3 so we will first look at the type parameters passed to C3.
TTT1 is used as the type parameter to C3<TT1, __, __> so it will infer the ```: class``` constraint.
TTT1 is used as the type parameter to C3<_, TT2, __> so it will infer the ```: IEnumerable<TTT1>``` constraint.
TTT1 is used as the type parameter to C3<__, __, TT3> so it will infer the ```: ICollection<TTT1>``` constraint.
TTT1 gains the explicit ```: IDictionary<long, TTT1>``` constraint

This would allow constraints to be inferred without having to be restated at each level of inheritance.

**Comment by HaloFour** [comment]
https://ericlippert.com/2012/03/09/why-not-automatically-infer-constraints/

**Comment by jnm2** [comment, marked as answer]
@TonyValenti You're not going to believe this. https://github.com/dotnet/csharplang/discussions/3810
